corehq/apps/enterprise/api/resources.py

```diff
@@ -138,12 +147,39 @@ def convert_datetime(cls, datetime_string):
         return time.isoformat()
 
 
-class ODataEnterpriseResource(ODataResource):
+class ODataEnterpriseReportResource(ODataResource):
     class Meta(ODataResource.Meta):
         authentication = EnterpriseODataAuthentication()
 
+    def get_list(self, request, **kwargs):
+        try:
+            response = super().get_list(request, **kwargs)
+        except ExpiredCacheException:
+            response = HttpResponseNotFound()
 
-class DomainResource(ODataEnterpriseResource):
+        return response
+
+    def get_object_list(self, request):
+        query_id = request.GET.get('query_id', None)
+        report = CacheableReport(self.get_report(request), query_id)
+        # HACK: Tastypie doesn't provide a hook to easily pass data to the paginator
+        # it is not ideal to perform a side effect within this get method, but the paginator
+        # assembles link data from the GET dictionary, so we are forced to modify the request somewhere
+        self._add_query_id_to_request(request, report.query_id)
+
+        return report.rows
+
+    def get_report(self, request):
+        raise NotImplementedError()
+
+    def _add_query_id_to_request(self, request, query_id):
+        if 'report' not in request.GET:
+            new_params = request.GET.copy()
+            new_params['query_id'] = query_id
+            request.GET = new_params
```

Comment:
~I'm not following how this works. Is it supposed to add `query_id` to the request parameters of future requests for this report? I think you'd need to redirect or set a cookie on the response or something along those lines. As far as I know modifying request GET parameters will have no direct effect on future requests.~

I think I get it. This value will be used by the paginator when constructing the response to the client. I didn't gather that from reading the big HACK comment above the first time. What do you think of adding a comment or docstring to this function to explain that a bit more clearly?

Reply:
I attempted to clarify the existing comment, rather than write a docstring for `_add_query_id_to_request`.  I felt that the question wasn't what `_add_query_id_to_request` was doing, but why we would want that functionality -- and that seemed to be a question for the caller, rather than for `_add_query_id_to_request`.